include/ginkgo/core/base/device.hpp

```diff
@@ -0,0 +1,140 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_DEVICE_HPP_
+#define GKO_PUBLIC_CORE_BASE_DEVICE_HPP_
+
+
+#include <array>
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+
+#include <ginkgo/config.hpp>
+
+
+namespace gko {
+
+
+class CudaExecutor;
+
+class HipExecutor;
+
+
+/**
+ * NvidiaDevice handles the number of executor on Nvidia devices and have the
+ * corresponding recursive_mutex.
+ */
+class NvidiaDevice {
+    friend class CudaExecutor;
+// If Hip is compiled for NVCC, give NvidiaDevice's permission to HipExecutor
+#if GINKGO_HIP_PLATFORM_NVCC
+    friend class HipExecutor;
+#endif
+
+private:
+    /**
+     * get_mutex gets the static mutex reference at i.
+     *
+     * @param i  index of mutex
+     *
+     * @return recursive_mutex reference
+     */
+    static std::recursive_mutex &get_mutex(int i)
+    {
+        static std::recursive_mutex mutex[max_devices];
+        return mutex[i];
+    }
+
+    /**
+     * get_num_execs gets the static num_execs reference at i.
+     *
+     * @param i  index of num_execs
+     *
+     * @return int reference
+     */
+    static int &get_num_execs(int i)
+    {
+        static int num_execs[max_devices];
+        return num_execs[i];
+    }
+
+    static constexpr int max_devices = 64;
+};
+
+
+/**
+ * AmdDevice handles the number of executor on Amd devices and have the
+ * corresponding recursive_mutex.
+ */
+class AmdDevice {
+// to avoid both GINKGO_HIP_PLATFORM_* zero
+// If Hip isn't compiled for NVCC, give AmdDevice's permission to HipExecutor
+#if !GINKGO_HIP_PLATFORM_NVCC
+    friend class HipExecutor;
+#endif
+
+private:
+    /**
+     * get_mutex gets the static mutex reference at i.
+     *
+     * @param i  index of mutex
+     *
+     * @return recursive_mutex reference
+     */
+    static std::recursive_mutex &get_mutex(int i)
+    {
+        static std::recursive_mutex mutex[max_devices];
+        return mutex[i];
+    }
+
+    /**
+     * get_num_execs gets the static num_execs reference at i.
+     *
+     * @param i  index of num_execs
+     *
+     * @return int reference
+     */
+    static int &get_num_execs(int i)
+    {
+        static int num_execs[max_devices];
```

Comment:
I'm a bit scared of this as well as the mutex declaration, technically, they could be instantiated anywhere depending on where this header is included first if I understand how this works properly. To make sure the correct thing always happens, maybe we should instead define this in `devices/machine_topology.cpp` or another file in this directory so that it's this definition is always available transitively everywhere (everything includes libginkgo_device)?